conjure-lib/src/main/java/com/palantir/conjure/java/lib/ImmutableByteArray.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.lib;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+/** An immutable {@code byte[]} wrapper. */
+public final class ImmutableByteArray {
+    private final byte[] safe;
+
+    /** Constructs a new {@link ImmutableByteArray} assuming the provided array is not held by any other class. */
+    private ImmutableByteArray(byte[] array) {
+        safe = array;
+    }
+
+    /** Returns a new read-only {@link ByteBuffer} backed by this byte array. */
+    public ByteBuffer getByteBuffer() {
+        return ByteBuffer.wrap(safe).asReadOnlyBuffer();
+    }
+
+    /** Returns a copy of this byte array. */
+    @JsonValue
```

Comment:
Should we move `@JsonValue` to `getByteBuffer` and avoid duplicating contents for serialization? Annotation may apply to the `safe` field, but I've not tested that type of usage.